spyder/plugins/editor/widgets/completion.py

```diff
@@ -0,0 +1,401 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright © Spyder Project Contributors
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
+
+"""Completion widget class"""
+
+# Standard library imports
+import sys
+
+# Third psrty imports
+from qtpy.QtCore import QPoint, Qt, Signal, Slot
+from qtpy.QtGui import QFontMetrics
+from qtpy.QtWidgets import (QAbstractItemView, QApplication, QListWidget,
+                            QListWidgetItem, QToolTip)
+
+# Local imports
+from spyder.utils import icon_manager as ima
+from spyder.plugins.completion.languageserver import CompletionItemKind
+from spyder.py3compat import to_text_string
+from spyder.widgets.helperwidgets import HTMLDelegate
+
+
+COMPLETION_ITEM_TEMPLATE = """
+<table width="{width}" height="{height}">
+    <tr>
+        <td valign="middle" style="color:{color}">
+            {completion}
+        </td>
+
+        <td valign="middle" align="right" float="right" style="color:{color}">
+            {type}
+        </td>
+    </tr>
+</table>
+"""
+DEFAULT_COMPLETION_ITEM_HEIGHT = 15
+DEFAULT_COMPLETION_ITEM_WIDTH = 250
+
+
+class CompletionWidget(QListWidget):
+    """Completion list widget"""
+
+    sig_show_completions = Signal(object)
+    sig_completion_hint = Signal(str, str, QPoint)
+
+    def __init__(self, parent, ancestor):
+        QListWidget.__init__(self, ancestor)
+        self.setWindowFlags(Qt.SubWindow | Qt.FramelessWindowHint)
+        self.textedit = parent
+        self.hide()
+        self.itemActivated.connect(self.item_selected)
+        self.currentRowChanged.connect(self.row_changed)
+        self.is_internal_console = False
+        self.completion_list = None
+        self.completion_position = None
+        self.automatic = False
+
+        # Setup item rendering
+        self.setItemDelegate(HTMLDelegate(self, margin=3))
+        self.setMinimumWidth(DEFAULT_COMPLETION_ITEM_WIDTH)
+
+        # Initial item height and width
+        fm = QFontMetrics(self.textedit.font())
+        self.item_height = fm.height()
+        self.item_width = self.width()
+
+    def setup_appearance(self, size, font):
+        self.resize(*size)
```

Comment:
Docstring?

Also, is this a setup (which is more like a one time thing only on creation) or is this more of a refresh (to be used periodically after creation)?

Reply:
I think it's only called in the `__init__` of the `TextEditBaseWidget`

Reply:
Ok, lets leave it then like it is :-)